flying-saucer-core/src/main/java/org/xhtmlrenderer/resource/HTMLResource.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * {{{ header & license
+ * Flying Saucer - Copyright (c) 2024
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License
+ * as published by the Free Software Foundation; either version 2.1
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+ * }}}
+ */
+package org.xhtmlrenderer.resource;
+
+import org.jsoup.Jsoup;
+import org.jsoup.helper.W3CDom;
+import org.jsoup.nodes.Document;
+import org.jsoup.parser.Parser;
+import org.xhtmlrenderer.util.XRLog;
+import org.xhtmlrenderer.util.XRRuntimeException;
+import java.io.*;
+import java.nio.charset.StandardCharsets;
+import java.net.URL;
+
+import javax.annotation.ParametersAreNonnullByDefault;
+
+/**
+ * HTMLResource uses JSoup to parse HTML.
+ */
+@ParametersAreNonnullByDefault
+public class HTMLResource extends AbstractResource {
+    private org.w3c.dom.Document document;
+
+    private HTMLResource(InputStream stream) {
+        super(stream);
+        try {
+            Document jsoupDoc = Jsoup.parse(stream, StandardCharsets.UTF_8.name(), "", Parser.htmlParser());
+            W3CDom w3cDom = new W3CDom();
+            this.document = w3cDom.fromJsoup(jsoupDoc);
+        } catch (IOException  e) {
+            XRLog.load(java.util.logging.Level.SEVERE, "Failed to parse and convert HTML document.", e);
+            throw new XRRuntimeException("Failed to parse and convert HTML document.", e);
+        }
+    }
+
+    public static HTMLResource load(URL source) {
+        try (InputStream stream = source.openStream()) {
+            return new HTMLResource(stream);
+        } catch (IOException e) {
+            XRLog.load(java.util.logging.Level.SEVERE, "Failed to load HTML from URL.", e);
+            throw new XRRuntimeException("Failed to load HTML from URL.", e);
+        }
+    }
+
+    public static HTMLResource load(InputStream stream) {
+        return new HTMLResource(stream);
+    }
+
+    public static HTMLResource load(Reader reader) {
+        try {
+            InputStream stream = convertReaderToInputStream(reader);
```

Comment:
Can jsoup read from a Reader directly? It would be nice to avoid the conversion.

Reply:
The main Jsoup class has no Reader method here: https://jsoup.org/apidocs/org/jsoup/Jsoup.html
However, there is https://jsoup.org/apidocs/org/jsoup/parser/Parser.html#parseInput(java.io.Reader,java.lang.String)